File: client-applications/savvy-web/src/app/services/saving.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Saving } from '../models/dashboard';

@Injectable({
  providedIn: 'root',
})
export class SavingService {
  private http = inject(HttpClient);
  private apiUrl = '/api/saving';

  createSaving(savingData: Partial<Saving>): Observable<Saving> {
    return this.http.post<Saving>(this.apiUrl, savingData);
  }

  getAllSavings(): Observable<Saving[]> {
    return this.http.get<Saving[]>(`${this.apiUrl}/all`);
  }

  getSavingById(id: number): Observable<Saving> {
    return this.http.get<Saving>(`${this.apiUrl}/${id}`);
  }

  updateSaving(id: number, savingData: Partial<Saving>): Observable<Saving> {
    return this.http.put<Saving>(`${this.apiUrl}/${id}`, savingData);
  }

  deleteSaving(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: domain-services/finance/savvy-service/src/main/java/com/ayman/distributed/savvy/services/saving/SavingService.java
package com.ayman.distributed.savvy.services.saving;

import com.ayman.distributed.savvy.model.dto.SavingDTO;
import com.ayman.distributed.savvy.model.entity.Saving;

import java.util.List;

public interface SavingService {
    Saving createSaving(SavingDTO savingDTO);
    List<Saving> getAllSavings();
    Saving getSavingById(Long id);
    Saving updateSaving(SavingDTO savingDTO, Long id);
    void deleteSaving(Long id);
}
